### PR TITLE
Add namespace

### PR DESCRIPTION
**Pull request by gboston**
I've added the ability to customize the jwt property that is placed on the fastify instance and allowed to customize function names on reply and request.

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: test.js [hunk 1]
```diff
@@ -50,6 +50,31 @@ test('register', function (t) {
     })
   })
 
+  t.test('Expose jwt methods namespaced', function (t) {
+    t.plan(7)
+
+    const fastify = Fastify()
+    fastify.register(jwt, { secret: 'test', namespace: 'security', jwtVerify: 'securityVerify', jwtSign: 'securitySign' })
```
Comment (mcollina) [inline test.js:57]: Can you register this twice with two difference namespaces in this test?

File: test.js [hunk 2]
```diff
@@ -1920,6 +1953,49 @@ test('custom response messages', function (t) {
...
+        url: '/verify',
+        headers: {
+          customauthheader: token
+        }
+      }).then(function (verifyResponse) {
+        t.is(verifyResponse.statusCode, 200)
+      })
+    })
+  })
+})
```
Comment (mcollina) [inline test.js:1997]: Would you mind to write this test using async/await instead? Thanks

File: jwt.js [hunk 3]
```diff
@@ -92,9 +91,25 @@ function fastifyJwt (fastify, options, next) {
...
-  fastify.decorateRequest('jwtVerify', requestVerify)
-  fastify.decorateReply('jwtSign', replySign)
+  }
+  if (options.namespace) {
+    if (!fastify.jwt) {
+      fastify.decorate('jwt', {})
+    }
+
+    if (fastify.jwt[options.namespace]) {
+      return next(new Error(`JWT namespace already used ${options.namespace}`))
```
Comment (Eomm) [inline jwt.js:101]:
```suggestion
      return next(new Error(`JWT namespace already used "${options.namespace}"`))
```

File: jwt.js [hunk 4]
```diff
@@ -92,9 +91,25 @@ function fastifyJwt (fastify, options, next) {
...
+    }
+    fastify.jwt[options.namespace] = jwtConfig
+  } else {
+    fastify.decorate('jwt', jwtConfig)
+  }
+
+  var jwtVerify = options.jwtVerify || 'jwtVerify'
```
Comment (Eomm) [inline jwt.js:108]:
```suggestion
  const jwtVerify = options.jwtVerify || 'jwtVerify'
```

If you merge master, the linter should drop the `var` keyword, so could you run `standard --fix`?

File: test.js [hunk 5]
```diff
@@ -50,6 +50,39 @@ test('register', function (t) {
     })
   })
 
+  t.test('Expose jwt methods namespaced', function (t) {
+    t.plan(14)
+
+    const fastify = Fastify()
+    fastify.register(jwt, { secret: 'test', namespace: 'security', jwtVerify: 'securityVerify', jwtSign: 'securitySign' })
+    fastify.register(jwt, { secret: 'hello', namespace: 'secure', jwtVerify: 'secureVerify', jwtSign: 'secureSign' })
```
Comment (Eomm) [inline test.js:58]: Could you add a 401 test too with these custom namespaces?
It is not clear to me what will happen: should the user trigger one token that verify to two authorization?

Could you update the readme accordingly?

**Review by mcollina — COMMENTED** [review]
could you add some docs about the new feature?

(I'm sorry this slipped through!)

File: test.js (hunk 5 above)
Comment (gboston) [reply test.js:58]: @Eomm Can you explain your question in more detail? I've added some documentation as requested.

**Comment by mcollina** [comment]
Can you add a test for the types? https://github.com/fastify/fastify-jwt/blob/master/jwt.test-d.ts

File: README.md [hunk 6]
```diff
@@ -449,6 +449,23 @@ fastify.register(require('fastify-jwt'), {
 })
 ```
 
+#### namespace options
+
+You can also define a namespace to make use of multiple JWT validators on the same routes. You can comebine this with custom names for jwtVerify and JwtSign.
```
Comment (Eomm) [inline README.md:454]:
```suggestion
You can also define a namespace to make use of multiple JWT validators on the same routes. You can combine this with custom names for `jwtVerify` and `jwtSign`.
```

File: test.js [hunk 7]
```diff
@@ -50,6 +50,39 @@ test('register', function (t) {
...
+      t.ok(fastify.jwt.security.verify)
+
+      t.ok(fastify.jwt.secure.decode)
+      t.ok(fastify.jwt.secure.options)
+      t.ok(fastify.jwt.secure.secret)
+      t.ok(fastify.jwt.secure.sign)
+      t.ok(fastify.jwt.secure.verify)
+    })
+
+    fastify.inject({
```
Comment (Eomm) [inline test.js:80]: this `inject` output is ignored (it is a promise without handling)

File: test.js (hunk 5 above)
Comment (Eomm) [reply test.js:58]: Nevermind, I was thinking of a bad usage of this feature

**Comment by mcollina** [comment]
This needs a rebase.

**Comment by mcollina** [comment]
ci is failing

**Comment by b-barry** [comment]
Hello @gboston,

Sorry to bother you but do you think you have enough time left to finish this PR?

Kind regards

**Comment by climba03003** [comment]
Closing as it stale for too long. Feel free to re-open it when you have free time.

**Comment by rymnc** [comment]
Can anyone re-run the CI so I can see where the error is?

**Pull request closed without merging**